tidy3d/components/time_modulation.py

```diff
@@ -0,0 +1,207 @@
+
+    """
+
+    @cached_property
+    @abstractmethod
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+
+
+class ContinuousWaveTimeModulation(AbstractTimeDependence):
+    """The modulation in time is in continuous oscillation.
+
+    Note
+    ----
+    .. math::
+
+        amp\\_time(t) = amplitude \\cdot \\
+                e^{i \\cdot phase - 2 \\pi i \\cdot freq0 \\cdot t}
+
+    Note
+    ----
+    The full space-time modulation is,
+    .. math::
+
+        amp(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+
+
+    Example
+    -------
+    >>> cw = ContinuousWaveTimeModulation(freq0=200e12, amplitude=1, phase=0)
+    """
+
+    freq0: pd.PositiveFloat = pd.Field(
+        ..., title="Modulation Frequency", description="Modulation frequency.", units=HERTZ
+    )
+
+    def amp_time(self, time: float) -> complex:
+        """Complex-valued source amplitude as a function of time."""
+
+        omega = 2 * np.pi * self.freq0
+        return self.amplitude * np.exp(-1j * omega * time + 1j * self.phase)
+
+    @cached_property
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+        return abs(self.amplitude)
+
+
+TimeModulationType = Union[ContinuousWaveTimeModulation]
+
+
+class AbstractSpaceModulation(ABC, Tidy3dBaseModel):
+    """Base class for modulation in space.
+
+    Note
+    ----
+    This class describes the 2nd term in the full space-time modulation below,
+    .. math::
+
+        amp(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+
+    """
+
+    @cached_property
+    @abstractmethod
+    def max_modulation(self) -> float:
+        """Estimated maximal modulation amplitude."""
+
+
+class SpaceModulation(AbstractSpaceModulation):
+    """The modulation profile in space with user-supplied spatial distribution of
+    amplitude and phase.
+
+    Note
+    ----
+    .. math::
+
+        amp\\_space(r) = amplitude(r) \\cdot \\e^{i \\cdot phase(r)}
+
+    The full space-time modulation is,
+    .. math::
+
+        amp(r, t) = \\Re[amp\\_time(t) \\cdot amp\\_space(r)]
+
+    Example
+    -------
+    >>> Nx, Ny, Nz = 10, 9, 8
+    >>> X = np.linspace(-1, 1, Nx)
+    >>> Y = np.linspace(-1, 1, Ny)
+    >>> Z = np.linspace(-1, 1, Nz)
+    >>> coords = dict(x=X, y=Y, z=Z)
+    >>> amp= SpatialDataArray(np.random.random((Nx, Ny, Nz)), coords=coords)
+    >>> phase= SpatialDataArray(np.random.random((Nx, Ny, Nz)), coords=coords)
+    >>> space = SpaceModulation(amplitude=amp, phase=phase)
+    """
+
+    amplitude: Union[float, SpatialDataArray] = pd.Field(
+        1,
+        title="Amplitude of modulation in space",
+        description="Amplitude of modulation that can vary spatially.",
+    )
+
+    phase: Union[float, SpatialDataArray] = pd.Field(
+        0,
+        title="Phase of modulation in space",
+        description="Phase of modulation that can vary spatially.",
```

Comment:
might be good to specify units of the phase (radians?) or at least whether this should be 2pi for a full cycle or 1.